File: sk-plugin/src/clientIntent.ts
// The intent one UI instance publishes to the plugin, plus this tab's
// stable identity. Kept out of src/pure/ because makeClientId() is
// deliberately non-deterministic (each browser tab must be distinguishable
// from every other) -- everything else here is a plain data shape.
//
// The wire counterpart lives in arbiter.cjs (server side); this is the
// TypeScript view of the exact object POSTed to the plugin's
// /plugins/<id>/intent route (not a Signal K path -- see config.ts).

import { SK_PLUGIN_INTENT_ENDPOINT } from './config';
import type { DrivePosition } from './pure/driveCommand';

/** Bow-thruster direction. 'off' (not 'neutral') -- a thruster coasts rather
 *  than sitting in a gear. Matches ThrusterCmdToSkString in the firmware. */
export type ThrusterDirection = 'port' | 'off' | 'stbd';

/** Which gate owns the thruster: the operator's buttons, or heading hold. */
export type ThrusterMode = 'manual' | 'hold';

export interface ClientIntent {
  /** Stable per-tab id; also the trailing path segment of this node. */
  clientId: string;
  /**
   * Increments on EVERY publish (including the 250 ms heartbeat), so no
   * change-detecting transport can ever dedupe a heartbeat and starve the
   * plugin's liveness view of this client.
   */
  seq: number;
  /** Increments once per operator "arm" tap. Edge-triggered by the arbiter. */
  armReq: number;
  /** Increments once per operator "disarm" tap. Edge-triggered; universal. */
  disarmReq: number;
  port: DrivePosition;
  stbd: DrivePosition;
  /** Bow thruster direction being asked for right now (manual mode only). */
  thruster: ThrusterDirection;
  /** Which thruster gate this client wants in force. */
  thrusterMode: ThrusterMode;
  /**
   * Commanded heading-hold TRIM in DEGREES, RELATIVE to the heading HH
   * captured when hold engaged. 0 = no trim (hold the captured heading). A
   * self-correcting level (the whole offset, not per-press events), clamped to
   * +/-MAX_TRIM_DEG, so a dropped or duplicated message can never desynchronise
   * this UI from the boat -- and, being relative, it needs no "not commanding"
   * sentinel (0 says it). See pure/trimOffset.ts and heading/heading_nudge.h.
   */
  trimDeg: number;
}

/**
 * A stable, unique-per-tab id. Two tabs (even on the same device) must get
 * different ids so the arbiter can tell them apart. If instances shared one
 * id (hence one $source), two disagreeing about ARMED would make the merged
 * value oscillate -- the separation is what prevents that. crypto.randomUUID()
 * where available, with a plain-random fallback for environments without it.
 */
export function makeClientId(): string {
  const c = (globalThis as { crypto?: Crypto }).crypto;
  if (c && typeof c.randomUUID === 'function') {
    return `ui-${c.randomUUID()}`;
  }
  return `ui-${Math.random().toString(36).slice(2)}-${Math.random()
    .toString(36)
    .slice(2)}`;
}

/** Sends one intent to the plugin. Injectable so tests can drive a real
 *  in-process arbiter without an HTTP server (mirrors how `fetchLoginStatus`
 *  is injected). Rejects on any non-2xx / network error so the caller can
 *  treat "commands not reaching the plugin" as a distinct state -- App.tsx
 *  turns the outcome into the Commands lamp via pure/intentStatus.ts. */
export type PostIntent = (intent: ClientIntent) => Promise<void>;

/**
 * A non-2xx answer to an intent POST. Carries the HTTP status so the caller
 * can tell "this browser is not allowed" (401/403) from "the plugin is not
 * running" (503) from a transport failure (no status at all).
 */
export class IntentPostError extends Error {
  readonly status: number;
  constructor(status: number) {
    super(`intent POST ${status}`);
    this.name = 'IntentPostError';
    this.status = status;
  }
}

/**
 * How long one intent POST may take. An intent is only relevant for the
 * 250 ms until the next one, so a request still open after this is already
 * stale -- and a server that has stopped answering must not be allowed to
 * queue an unbounded backlog of obsolete commands behind the browser's
 * per-origin connection limit, with a STOP somewhere at the back of it.
 */
export const INTENT_POST_TIMEOUT_MS = 2000;

/**
 * Production transport: POST the intent to the plugin's own route. Relative +
 * same-origin (the app is served BY the SK server), `credentials: 'include'`
 * so it carries the same session cookie the WebSocket does -- meaning a
 * secured server authorizes (or 401-rejects) it exactly as it would a delta
 * write. Nothing is written to the Signal K data tree.
 */
export const postIntent: PostIntent = async (intent) => {
  // AbortSignal.timeout is missing on some older WebViews; without it the
  // request simply has no client-side deadline, as before.
  const signal =
    typeof AbortSignal !== 'undefined' && typeof AbortSignal.timeout === 'function'
      ? AbortSignal.timeout(INTENT_POST_TIMEOUT_MS)
      : undefined;
  const res = await fetch(SK_PLUGIN_INTENT_ENDPOINT, {
    method: 'POST',
    credentials: 'include',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(intent),
    signal,
  });
  if (!res.ok) throw new IntentPostError(res.status);
};

File: android/core/src/main/kotlin/io/github/kegustafsson/driveremote/core/ClientIntent.kt
package io.github.kegustafsson.driveremote.core

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put

/**
 * What one station POSTs to the arbiter. Port of
 * `sk-plugin/src/clientIntent.ts`; the wire counterpart is `arbiter.cjs`.
 *
 * This is an INTENT, not a command. The station says what it wants; the
 * server-side arbiter decides who holds the single arm token and is the sole
 * writer of the `plugin.*` paths. That is what lets a phone, a tablet and a
 * browser all be open at once without their armed states fighting -- and it is
 * why this app, like the browser UI, never writes a Signal K path itself.
 */

/** Bow-thruster direction. 'off', not 'neutral' -- a thruster coasts rather than sitting in gear. */
enum class ThrusterDirection(val wire: String) {
  PORT("port"),
  OFF("off"),
  STBD("stbd"),
}

/** Which gate owns the thruster: the operator's buttons, or heading hold. */
enum class ThrusterMode(val wire: String) {
  MANUAL("manual"),
  HOLD("hold"),
}

data class ClientIntent(
  /** Stable per-installation id. Two stations must never share one. */
  val clientId: String,
  /**
   * Per-PROCESS id, regenerated on every launch -- unlike [clientId], which is
   * persisted because the Signal K access token is issued to that device id.
   *
   * ORDERED, not merely unique: it lets the arbiter say which of two sessions is
   * newer and close the older one permanently. It is what tells a restart from a
   * replay. [seq] and the
   * counters below live in memory and begin again at 0 on relaunch, so a
   * process restarted inside the arbiter's stale timeout sends packets with a
   * LOWER seq than the record it left behind -- indistinguishable from a
   * delayed heartbeat, and discarded by the ordering gate. That silently
   * swallowed the new session's STOP. A changed session id says "new session"
   * outright rather than asking the arbiter to guess from the counters.
   */
  val session: Long,
  /**
   * Increments on EVERY send, including the 250 ms heartbeat, so no
   * change-detecting transport can dedupe a heartbeat and starve the arbiter's
   * liveness view of this client. The arbiter also enforces strict ordering on
   * it: an intent whose seq is not greater than the last accepted one is
   * discarded whole, so a delayed heartbeat cannot revert live commands.
   */
  val seq: Long,
  /** Increments once per operator ARM tap. Edge-triggered by the arbiter. */
  val armReq: Long,
  /** Increments once per operator DISARM tap. Edge-triggered; universal. */
  val disarmReq: Long,
  val port: DrivePosition,
  val stbd: DrivePosition,
  /** Thruster direction wanted right now (manual mode only). */
  val thruster: ThrusterDirection,
  val thrusterMode: ThrusterMode,
  /**
   * Heading-hold TRIM in degrees, RELATIVE to the heading HH captured when hold
   * engaged. 0 = no trim. Published as a self-correcting LEVEL rather than
   * per-press events: an event has to arrive exactly once to stay correct,
   * while a level is self-correcting over a lossy link and survives a reboot.
   */
  val trimDeg: Double,
) {
  /**
   * The exact JSON the arbiter expects. Field names are the wire contract with
   * `arbiter.cjs` -- kept in lockstep with `clientIntent.ts`.
   */
  fun toJson(): String =
    buildJsonObject {
        put("clientId", clientId)
        put("session", session)
        put("seq", seq)
        put("armReq", armReq)
        put("disarmReq", disarmReq)
        put("port", port.wire)
        put("stbd", stbd.wire)
        put("thruster", thruster.wire)
        put("thrusterMode", thrusterMode.wire)
        put("trimDeg", trimDeg)
      }
      .toString()

  companion object {
    /**
     * The intent a station sends when it is not commanding anything: disarmed,
     * both drives neutral, thruster off, no trim. This is the value the app
     * falls back to whenever it loses the right (or the ability) to command --
     * SAFETY.md: fail to the safe value, never to the last commanded one.
     */
    fun safe(
      clientId: String,
      session: Long,
      seq: Long,
      armReq: Long,
      disarmReq: Long,
    ): ClientIntent =
      ClientIntent(
        clientId = clientId,
        session = session,
        seq = seq,
        armReq = armReq,
        disarmReq = disarmReq,
        port = DrivePosition.NEUTRAL,
        stbd = DrivePosition.NEUTRAL,
        thruster = ThrusterDirection.OFF,
        thrusterMode = ThrusterMode.MANUAL,
        trimDeg = 0.0,
      )
  }
}

/**
 * Parses the arbiter's acknowledgement. The body is only `{ok:true}` today and
 * the app deliberately does not depend on it -- a station learns who holds the
 * token from the `activeClient` path it already subscribes to, never from this
 * reply. Parsed anyway so a malformed body is distinguishable from a transport
 * failure in logs.
 */
fun parseIntentAck(body: String): Boolean =
  runCatching {
      val obj = Json.parseToJsonElement(body) as? JsonObject ?: return false
      // booleanOrNull, not content == "true": the latter would also accept the
      // STRING "true", which is not what the route sends and not what we mean.
      (obj["ok"] as? JsonPrimitive)?.booleanOrNull == true
    }
    .getOrDefault(false)
